File: acwing/算法基础课/第四讲/868.筛质数/868.筛质数.py
N = int(1e6 + 10)
primes = [0] * N
st = [0] * N
cnt = 0

n = int(input())

for i in range(2, n + 1):
    if st[i] == 0:
        primes[cnt] = i
        cnt += 1
    j = 0
    while primes[j] <= n // i:
        st[primes[j] * i] = 1
        if i % primes[j] == 0:
            break
        j += 1

print(cnt)

File: acwing/算法基础课/第四讲/868.筛质数/868.筛质数.cpp
#include <iostream>
#include <cstring>
#include <algorithm>

using namespace std;

const int N = 1e6 + 10;

int primes[N], cnt, st[N];

int main()
{
    int n;
    cin >> n;
    for (int i = 2; i <= n; i++)
    {
        if (!st[i])
        {
            st[i] = 1;
            primes[cnt++] = i;
        }
        for (int j = 0; primes[j] <= n / i; j++)
        {
            st[primes[j] * i] = 1;
            if (i % primes[j] == 0)
                break;
        }
    }
    cout << cnt << endl;
    return 0;
}